File: dsa-05-main/python/day19/BellmanFord.py


class BellmanFord:


	def __init__(self, vCount):
		self.INF = 999
		self.vertCount = vCount
		self.edgeCount = 0
		self.mat = []
		self.edges = []
		for i in range(self.vertCount):
			self.mat.append([self.INF] * self.vertCount)
	

	def accept(self):
		self.edgeCount = int(input("Enter number of edges: "))
		print(f"Enter {self.edgeCount} edges -- src dest weight:")
		for i in range(self.edgeCount):
			(src_str, dest_str, wt_str) = input().split()
			(src,dest,wt) = (int(src_str), int(dest_str), int(wt_str))
			self.mat[src][dest] = wt
			self.mat[dest][src] = wt # for undirected graph src -> dest => dest -> src 
			self.edges.append((src,dest,wt))


	def display(self):
		for i in range(self.vertCount):
			print(f"v{i}", end='')
			for j in range(self.vertCount):
				if(self.mat[i][j] == self.INF):
					print("\t#", end='')
				else:
					print("\t{}".format(self.mat[i][j]), end='')
			print()


	def bellmanFord(self, start):
		# init dist of start as 0 and others as INF
		dist = [self.INF] * self.vertCount
		dist[start] = 0

		# repeat for V-1 times
		for i in range(1, self.vertCount):
			# for all edges
			for (src,dest,weight) in self.edges:
				# update weights of all dest
				if(dist[src] != self.INF and dist[src] + weight < dist[dest]):
					dist[dest] = dist[src] + weight
		
		#  check for -ve cycle
		for (src,dest,weight) in self.edges:
			# update weights of all dest
			if(dist[src] != self.INF and dist[src] + weight < dist[dest]):
				raise Exception("Graph contains -ve cycle.")
	
		return dist


if __name__=="__main__":
	g = BellmanFord(5)
	g.accept()
	g.display()

	dist = g.bellmanFord(0)
	print("dist from 0 - " + str(dist))


"""
7
3 4 3
2 4 3
2 3 4
2 1 -2
1 3 -1
0 2 5
0 1 6 
"""

File: dsa-05-main/cpp/day19/BellmanFord.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = 999;

class Edge {
	private:
		int src, dest, weight;
	public:
	Edge(int src, int dest, int weight) {
		this->src = src;
		this->dest = dest;
		this->weight = weight;
	}
	int getWeight() {
		return weight;
	}
	void display() {
		cout << "Edge [src=" << src << ", dest=" << dest << ", weight=" << weight << "]" << endl;
	}
	static bool compEdgeWeight(const Edge& e1, const Edge& e2) {
		return e1.weight < e2.weight;
	}
	friend class BellmanFord;
};

class BellmanFord {
private:
	int **mat;
	int vertCount, edgeCount;
	vector<Edge> edges;
public:
	BellmanFord(int vCount)
	{
		vertCount = vCount;
		edgeCount = 0;
		mat = new int *[vertCount];
		for (int i = 0; i < vertCount; i++)
		{
			mat[i] = new int[vertCount];
			for (int j = 0; j < vertCount; j++)
				mat[i][j] = 0;
		}
	}
	~BellmanFord()
	{
		for (int i = 0; i < vertCount; i++)
			delete[] mat[i];
		delete[] mat;
		mat = NULL;
		vertCount = 0;
		edgeCount = 0;
	}

	void accept()
	{
		cout << "Enter number of edges: ";
		cin >> edgeCount;
		cout << "Enter " << edgeCount << " edges -- src dest weight:" << endl;
		for (int i = 0; i < edgeCount; i++)
		{
			int src, dest, wt;
			cin >> src >> dest >> wt;
			mat[src][dest] = wt;
			Edge e(src, dest, wt);
			edges.push_back(e);
		}
	}

	void display()
	{
		for (int i = 0; i < vertCount; i++)
		{
			cout << "v" << i;
			for (int j = 0; j < vertCount; j++)
				cout << "\t" << mat[i][j];
			cout << endl;
		}
	}

	int* bellmanFord(int start) {
		int* dist = new int[vertCount];
		// init dist of start as 0 and others as INF
		for (int i = 0; i < vertCount; i++)
			dist[i] = INF;
		dist[start] = 0;

		// repeat for V-1 times
		for(int i=1; i<vertCount; i++) {
			// for all edges
			for(int j=0; j<edgeCount; j++) {
				Edge e = edges[j];
				// update weights of all dest
				if(dist[e.src] != INF && dist[e.src] + e.weight < dist[e.dest])
					dist[e.dest] = dist[e.src] + e.weight;
			}
		}
		
		// check for -ve cycle
		for(int j=0; j<edgeCount; j++) {
			Edge e = edges[j];
			// update weights of all dest
			if(dist[e.src] != INF && dist[e.src] + e.weight < dist[e.dest])
				throw "Graph contains -ve cycle.";
		}
		
		return dist;
	}
};

int main() {
	int vertCount = 5;
	BellmanFord g(vertCount);
	g.accept();
	g.display();
	
	int* dist;
	dist = g.bellmanFord(0);
	for (int i = 0; i < vertCount; i++)
		cout << "distance of vertex " << i << " from vertex 0 = " << dist[i] << endl;
	
	delete[] dist;
	return 0;
}



/*
7
3 4 3
2 4 3
2 3 4
2 1 -2
1 3 -1
0 2 5
0 1 6 
 */

